test/jdk/java/awt/print/PrinterJob/Margins.java

```diff
@@ -47,28 +49,33 @@
 
     public static void main(String args[]) throws Exception {
         Robot robot = new Robot();
-        Thread t = new Thread (() -> {
-            robot.waitForIdle();
-            robot.delay(5000);
-            robot.keyPress(KeyEvent.VK_ENTER);
-            robot.keyRelease(KeyEvent.VK_ENTER);
-            robot.waitForIdle();
-            robot.delay(5000);
-            robot.keyPress(KeyEvent.VK_ENTER);
-            robot.keyRelease(KeyEvent.VK_ENTER);
-            robot.waitForIdle();
-            robot.delay(5000);
-            robot.keyPress(KeyEvent.VK_ENTER);
-            robot.keyRelease(KeyEvent.VK_ENTER);
-        });
-
         PrinterJob job = PrinterJob.getPrinterJob();
+        if (job.getPrintService() == null) {
+            System.out.println("No printers. Test cannot continue");
+            return;
+        }
+        PrintService psrv = PrintServiceLookup.lookupDefaultPrintService();
+        System.out.println("PrintService " + psrv.getName());
+        if (psrv.getName().contains("Print to PDF") ||
+            psrv.getName().contains("OneNote") ||
+            psrv.getName().contains("XPS Document Writer")) {
+            System.out.println("Please use real printer.");
+            System.out.println("Please avoid Microsoft Print to PDF or OneNote printer or XPS Writer");
+            return;
+        }
+
         PageFormat pageFormat = job.defaultPage();
         Paper paper = pageFormat.getPaper();
         double wid = paper.getWidth();
         double hgt = paper.getHeight();
         paper.setImageableArea(0, -10, wid, hgt);
-        t.start();
+
+        Thread t1 = new Thread(() -> {
+            robot.delay(2000);
+            robot.keyPress(KeyEvent.VK_ENTER);
+            robot.keyRelease(KeyEvent.VK_ENTER);
+        });
+        t1.start();
```

Comment:
Don't you need to wait until the t1(and others) thread completes? You call start() but the actual execution may not start till the end of the test.